journal.py

```diff
@@ -92,45 +103,94 @@ def home(request):
+                                'title': title,
+                                'text': text,
+                                'id': new_entry.id,
+                                'time': new_entry.created.strftime('%b. %d, %Y')}),
+                                content_type=b'application/json')
         else:
             return {'title': title, 'text': text}
     else:
         return {'title': '', 'text': ''}
 
 
+@view_config(route_name='add', renderer='templates/add.jinja2')
+def add(request):
+    return do_add(request, True)
+
+
+@view_config(route_name='ajax_add', renderer='templates/ajax_add.jinja2')
+def ajax_add(request):
+    return do_add(request, False)
+
+
+@view_config(route_name='added', renderer='templates/added.jinja2')
+def added(request):
+    title = request.matchdict['title']
+    entry = Entry.get(title=title)
+    return {'entry': entry}
+
+
 @view_config(route_name='detail', renderer='templates/detail.jinja2')
 def detail(request):
     entry_id = request.matchdict['entry_id']
-    title, text, time = Entry.get_entry(entry_id)
-    text = markdown.markdown(text, extensions=['codehilite', 'fenced_code'])
-    return {'title': title, 'text': text, 'id': entry_id, 'time': time}
+    entry = Entry.get(entry_id)
+    text = markdown.markdown(entry.text,
+                             extensions=['codehilite', 'fenced_code'])
+    return {'title': entry.title,
+            'text': text,
+            'id': entry_id,
+            'time': entry.created.strftime('%b. %d, %Y')}
 
 
-@view_config(route_name='edit', renderer='templates/edit.jinja2')
-def edit(request):
+def do_edit(request, redirect):
     entry_id = request.matchdict['entry_id']
     if request.method == 'GET':
-        title, text, time = Entry.get_entry(entry_id)
-        return {'title': title, 'text': text, 'id': entry_id}
-
-    if request.method == 'POST':
+        entry = Entry.get(entry_id)
+        return {'title': entry.title, 'text': entry.text, 'id': entry_id}
+    elif request.method == 'POST':
         title = request.params.get('title')
         text = request.params.get('text')
         if not (title == "" or text == ""):
             Entry.edit_entry(entry_id, title, text)
-            return HTTPFound(request.route_url('home'))
+            if redirect:
+                return HTTPFound(request.route_url(
+                                 'detail', entry_id=entry_id))
+            else:
+                entry = Entry.get(entry_id)
+                return Response(body=json.dumps({
+                                'title': entry.title,
+                                'text': markdown.markdown(
+                                 entry.text, extensions=['codehilite',
+                                                         'fenced_code']),
+                                'id': entry.id,
+                                'time': entry.created.strftime('%b. %d, %Y')}),
+                                content_type=b'application/json')
```

Comment:
I'm not sure that pushing the complexity of this logic down into these `do_*` methods is buying you much in terms of DRY.  You can achieve the same effect by using view config predicates to filter which requests go where.